### PR TITLE
Filter based on name and tag

### PR DESCRIPTION
**Pull request by malwaregarry**
### Added
* #38 : filtering based on tags
* #39 : filtering based on name
* Clearing of specified filters which were applied
* Clearing of all filters applied
### Removed
* FindCommand
* ListCommand

Tests will be in another PR

### CODE REVIEW

**Review by craeyeons — COMMENTED** [review]
LGTM.

File: src/main/java/seedu/address/model/person/TagMatchesQueryPredicate.java [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+        Set<Tag> intersection = new HashSet<>(queries);
+        intersection.retainAll(person.getTags());
+        return !intersection.isEmpty();
+    }
+
+    @Override
+    public boolean equals(Object other) {
+        return other == this // short circuit if same object
+                || (other instanceof TagMatchesQueryPredicate // instanceof handles nulls
+                        && queries.containsAll(((TagMatchesQueryPredicate) other).queries)); // state check
```
[the remark this answers is not in the document]
Comment (malwaregarry) [reply src/main/java/seedu/address/model/person/TagMatchesQueryPredicate.java:31]: Hi, I don't see where the inconsistency is here

File: src/main/java/seedu/address/model/person/NameContainsKeywordsPredicate.java [hunk 2]
```diff
@@ -25,7 +34,12 @@ public boolean test(Person person) {
     public boolean equals(Object other) {
         return other == this // short circuit if same object
                 || (other instanceof NameContainsKeywordsPredicate // instanceof handles nulls
-                && keywords.equals(((NameContainsKeywordsPredicate) other).keywords)); // state check
+                        && keywords.equals(((NameContainsKeywordsPredicate) other).keywords)); // state check
```
[the remark this answers is not in the document]
Comment (malwaregarry) [reply src/main/java/seedu/address/model/person/NameContainsKeywordsPredicate.java:37]: Fixed in [29000a3](https://github.com/AY2223S1-CS2103T-T13-2/tp/pull/43/commits/29000a38f7ae7364b4c854320504cfffde1d3372)

File: src/main/java/seedu/address/model/person/TagMatchesQueryPredicate.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (malwaregarry) [reply src/main/java/seedu/address/model/person/TagMatchesQueryPredicate.java:31]: Fixed in [29000a3](https://github.com/AY2223S1-CS2103T-T13-2/tp/pull/43/commits/29000a38f7ae7364b4c854320504cfffde1d3372)

File: src/main/java/seedu/address/logic/parser/ParserUtil.java [hunk 3]
```diff
@@ -137,4 +142,52 @@ public static Set<Tag> parseTags(Collection<String> tags) throws ParseException
...
+        requireNonNull(tags);
+        Set<Tag> tagSetQuery = parseTags(tags);
+        return new TagMatchesQueryPredicate(tagSetQuery);
+    }
+
+    /**
+     * Parses {@code String keywords} separated by commas into a {@code List<String>}.
+     *
+     * @param keywords Comma separated strings.
+     * @return A list of keywords that were separated.
```
[the remark this answers is not in the document]
Comment (malwaregarry) [reply src/main/java/seedu/address/logic/parser/ParserUtil.java]: Fixed in [54e0a28](https://github.com/AY2223S1-CS2103T-T13-2/tp/pull/43/commits/54e0a28efe7dda79273a9b036eca5ebcd8dc7d74)

**Review by Polygonalr — COMMENTED** [review]
LGTM aside from cosmetic issues

File: src/main/java/seedu/address/logic/commands/FilterCommand.java [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+ * Filters contacts in address book whose name contains the argument keywords.
+ * Keyword matching is case insensitive.
+ */
+public class FilterCommand extends Command {
+
+    public static final String COMMAND_WORD = "filter";
+
+    public static final String MESSAGE_USAGE = COMMAND_WORD + ": Filters all contacts whose names contain any of "
+            + "the specified keywords (case-insensitive) and displays them as a list with index numbers.\n"
+            + "Parameters: " + PREFIX_NAME + "NAME ...\n" + "Example: " + COMMAND_WORD + PREFIX_NAME + "alice";
```
[the remark this answers is not in the document]
Comment (malwaregarry) [reply src/main/java/seedu/address/logic/commands/FilterCommand.java:20]: Good catch! Can't believe I didn't spot all these

**Comment by kelvinou01** [comment]
LGTM!

**Pull request merged**